### PR TITLE
ADCM-6155: Implement missing methods for ServiceNode

### PR DESCRIPTION
**Pull request by Starovoitov**

### CODE REVIEW

File: adcm_aio_client/core/objects/cm.py [hunk 1]
```diff
@@ -192,6 +192,19 @@ def components(self: Self) -> "ComponentsNode":
...
+        return ADCMEntityStatus(response.as_dict()["status"])
+
+    async def delete(self: Self) -> None:
+        await self._requester.delete(*self._path)
+
+    async def turn_maintenance_mode_on(self: Self) -> None:
+        await self._requester.post(*self._path, "maintenance-mode", data={"maintenanceMode": MaintenanceModeStatus.ON})
+
+    async def turn_maintenance_mode_off(self: Self) -> None:
+        await self._requester.post(*self._path, "maintenanceMode", data={"maintenanceMode": MaintenanceModeStatus.OFF})
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:206]: how you can do any of these operations on Node?

it's object's methods at  best (and most of them are implemented, I believe)

---

`add` here is missing

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py (hunk 1 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py:206]: What is Node supposed for then? add is not listed in requirements here https://wiki.adsw.io/books/adcm/page/adcm-aio-client#bkmrk-host-attributes%3A-id%3B-1

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py [hunk 2]
```diff
@@ -188,9 +199,28 @@ def components(self: Self) -> "ComponentsNode":
...
+    def _get_id_by_name(self: Self, name: str, items: list[dict]) -> int:
+        for item in items:
+            if item["name"] == name:
+                return item["id"]
+        raise NotFoundError
+
+    async def add(self: Self, name: str) -> Service:
+        prototypes = (await self._requester.get("prototypes")).as_list()
+        prototype_id = self._get_id_by_name(name, prototypes)
+        response = await self._requester.post("services", data={"prototypeId": prototype_id, "name": name})
```
[the remark this answers is not in the document]
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: image
this one should be correct

File: adcm_aio_client/core/objects/cm.py [hunk 3]
```diff
@@ -188,9 +199,28 @@ def components(self: Self) -> "ComponentsNode":
...
     class_type = Service
 
+    def _get_id_by_name(self: Self, name: str, items: list[dict]) -> int:
+        for item in items:
+            if item["name"] == name:
+                return item["id"]
+        raise NotFoundError
+
+    async def add(self: Self, name: str) -> Service:
+        prototypes = (await self._requester.get("prototypes")).as_list()
```
[the remark this answers is not in the document]
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: done

File: adcm_aio_client/core/objects/cm.py [hunk 4]
```diff
@@ -188,9 +199,28 @@ def components(self: Self) -> "ComponentsNode":
 class ServicesNode(PaginatedChildAccessor[Cluster, Service, None]):
     class_type = Service
 
+    def _get_id_by_name(self: Self, name: str, items: list[dict]) -> int:
+        for item in items:
+            if item["name"] == name:
+                return item["id"]
+        raise NotFoundError
+
+    async def add(self: Self, name: str) -> Service:
```
[the remark this answers is not in the document]
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: changed

File: adcm_aio_client/core/objects/_common.py [hunk 5]
```diff
@@ -47,3 +47,32 @@ def config_groups(self: Self) -> ...: ...
 class WithActionHostGroups(WithProtectedRequester, AwareOfOwnPath):
     @cached_property
     def action_host_groups(self: Self) -> ...: ...
+
+
+class WithMaintenanceMode(WithProtectedRequester, AwareOfOwnPath):
+    class _MaintenanceMode[T]:
```
[the remark this answers is not in the document]
Comment (Starovoitov) [reply adcm_aio_client/core/objects/_common.py]: changed as it in https://github.com/arenadata/adcm-aio-client/pull/30/files

File: adcm_aio_client/core/objects/cm.py [hunk 6]
```diff
@@ -225,10 +237,53 @@ def cluster(self: Self) -> Cluster:
...
         return ComponentsNode(parent=self, path=(*self.get_own_path(), "components"), requester=self._requester)
 
+    @property
+    def license(self: Self) -> License:
+        return License(self._requester, self._data)
+
+    @async_cached_property
+    async def maintenance_mode(self: Self) -> MaintenanceMode:
+        response = await self._requester.get(*self.get_own_path())
+        self._data = response.as_dict()
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:247]: I see no reason for re-requesting object for maintenance mode sake

see comment in your's another MR on this matter

**Review by a-alferov — APPROVED** [review]

**Review by Sealwing — APPROVED** [review]

**Pull request merged**